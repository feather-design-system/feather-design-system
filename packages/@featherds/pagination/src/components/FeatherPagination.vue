<template>
  <div class="feather-pagination">
    <div class="per-page-text">
      {{ perPageText }}
    </div>
    <FeatherSelect
      :label="perPageText"
      :modelValue="_pageSize"
      @update:modelValue="updatePageSize"
      :options="_pageSizes"
      inline
      class="page-size-select"
    />
    <div class="range-text">
      {{ rangeText }}
    </div>
    <nav :aria-label="paginationText">
      <ul>
        <li>
          <FeatherButton
            class="action"
            as-anchor
            href="#"
            ref="first"
            :icon="firstText"
            :disabled="disableBackward"
            @click.prevent="first"
            data-ref-id="feather-pagination-first-button"
          >
            <FeatherIcon :icon="firstIcon" />
          </FeatherButton>
        </li>
        <li>
          <FeatherButton
            class="action"
            as-anchor
            href="#"
            :icon="previousText"
            :disabled="disableBackward"
            @click.prevent="previous"
            data-ref-id="feather-pagination-previous-button"
          >
            <FeatherIcon :icon="previousIcon" />
          </FeatherButton>
        </li>
        <li>
          <FeatherButton
            class="action"
            as-anchor
            href="#"
            :icon="nextText"
            :disabled="disableForward"
            @click.prevent="next"
            data-ref-id="feather-pagination-next-button"
          >
            <FeatherIcon :icon="nextIcon" />
          </FeatherButton>
          <FeatherButton
            class="action"
            as-anchor
            href="#"
            :icon="lastText"
            :disabled="disableForward"
            @click.prevent="last"
            data-ref-id="feather-pagination-last-button"
          >
            <FeatherIcon :icon="lastIcon" />
          </FeatherButton>
        </li>
      </ul>
    </nav>
  </div>
</template>
<script>
import { FeatherSelect } from "@featherds/select";
import { FeatherButton } from "@featherds/button";
import { FeatherIcon } from "@featherds/icon";
import ChevronLeft from "@featherds/icon/navigation/ChevronLeft";
import ChevronRight from "@featherds/icon/navigation/ChevronRight";
import FirstPage from "@featherds/icon/navigation/FirstPage";
import LastPage from "@featherds/icon/navigation/LastPage";
import { markRaw } from "vue";

const LABELS = {
  rowsPerPage: "Rows per page",
  range: "${start} - ${end} of ${total}",
  first: "Go to first page",
  last: "Go to last page",
  next: "Go to next page",
  previous: "Go to previous page",
  paginationLabel: "Pagination controls",
};
export default {
  emits: ["update:modelValue", "update:pageSize"],
  model: {
    prop: "modelValue",
    event: "update:modelValue",
  },
  props: {
    labels: {
      type: Object,
      default: () => {
        return LABELS;
      },
    },
    modelValue: {
      type: Number,
      required: true,
    },
    total: {
      type: Number,
      required: true,
    },
    pageSizes: {
      type: Array,
      default: () => [10, 20, 50],
    },
    pageSize: {
      type: Number,
      required: true,
      default: 10,
    },
  },
  computed: {
    firstText() {
      return this.labels.first ? this.labels.first : LABELS.first;
    },
    firstIcon() {
      return markRaw(FirstPage);
    },
    lastText() {
      return this.labels.last ? this.labels.last : LABELS.last;
    },
    lastIcon() {
      return markRaw(LastPage);
    },
    nextText() {
      return this.labels.next ? this.labels.next : LABELS.next;
    },
    nextIcon() {
      return markRaw(ChevronRight);
    },
    previousText() {
      return this.labels.previous ? this.labels.previous : LABELS.previous;
    },
    previousIcon() {
      return markRaw(ChevronLeft);
    },
    perPageText() {
      return this.labels.rowsPerPage
        ? this.labels.rowsPerPage
        : LABELS.rowsPerPage;
    },
    paginationText() {
      return this.labels.paginationLabel
        ? this.labels.paginationLabel
        : LABELS.paginationLabel;
    },
    lastPage() {
      return Math.ceil(this.total / this.pageSize);
    },
    disableForward() {
      return this.modelValue >= this.lastPage;
    },
    disableBackward() {
      return this.modelValue <= 1;
    },
    rangeText() {
      let result = this.labels.range ? this.labels.range : LABELS.range;
      result = result.replace("${total}", this.total);
      const start = this.modelValue * this.pageSize - this.pageSize + 1;
      if (start < 0 || start > this.total) {
        return "ERROR calculating start";
      }
      result = result.replace("${start}", start);
      let end = start + this.pageSize - 1;
      if (end > this.total) {
        end = this.total;
      }
      result = result.replace("${end}", end);
      return result;
    },
    _pageSize() {
      return { _text: this.pageSize };
    },
    _pageSizes() {
      return this.pageSizes.map((x) => {
        return {
          _text: x,
        };
      });
    },
  },
  methods: {
    updatePageSize(e) {
      this.$emit("update:pageSize", parseInt(e._text, 10));
      this.first();
    },
    first() {
      this.$emit("update:modelValue", 1);
    },
    last() {
      this.$emit("update:modelValue", this.lastPage);
    },
    next() {
      if (this.modelValue + 1 >= this.lastPage) {
        this.$emit("update:modelValue", this.lastPage);
        return;
      }
      this.$emit("update:modelValue", this.modelValue + 1);
    },
    previous() {
      if (this.modelValue - 1 <= 1) {
        this.$emit("update:modelValue", 1);
        return;
      }
      this.$emit("update:modelValue", this.modelValue - 1);
    },
  },
  components: {
    FeatherSelect,
    FeatherButton,
    FeatherIcon,
  },
};
</script>

<style lang="scss" scoped>
@import "~@featherds/styles/themes/variables";
.feather-pagination {
  display: flex;
  border: 1px solid var($border-on-surface);
  color: var($secondary-text-on-surface);
  height: 56px;
  align-items: center;
  justify-content: flex-end;
  margin-bottom: $spacing-l;
  .action {
    margin: 0 $spacing-m;
  }
  ul {
    padding: 0;
    margin: 0;
    li {
      list-style: none;
      display: inline-block;
    }
  }
}
.page-size-select {
  width: 76px;
  margin-right: 38px;
  margin-left: 8px;
  display: flex;
}
.range-text {
  margin-right: 50px;
  min-width: 110px;
}
</style>
