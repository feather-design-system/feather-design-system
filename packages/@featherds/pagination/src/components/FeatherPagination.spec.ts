import FeatherPagination from "./FeatherPagination.vue";
import { shallowMount } from "@vue/test-utils";
import { getCalls } from "@featherds/utils/test/calls";
import { expect, describe, it } from "vitest";

const getWrapper = function (options: Record<string, unknown> = {}) {
  const props = (options.props as Object) || {};
  options.props = {
    pageSize: 10,
    ...props,
  };
  return shallowMount(FeatherPagination, options);
};

describe("FeatherPagination.vue", () => {
  it("should disable first and previous buttons on first page", () => {
    const wrapper = getWrapper({
      props: {
        modelValue: 1,
        total: 100,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should disable last and next buttons on last page", () => {
    const wrapper = getWrapper({
      props: {
        modelValue: 2,
        total: 20,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should emit update event when pageSize is changed", () => {
    const wrapper = getWrapper({
      props: {
        modelValue: 1,
        total: 100,
      },
    });
    wrapper.vm.updatePageSize({ _text: "20" });
    expect(getCalls<[number]>(wrapper, "update:pageSize")[0][0]).toBe(20);
    expect(getCalls<[number]>(wrapper, "update:modelValue")[0][0]).toBe(1);
  });
  it("should set page to 1 when first is clicked", () => {
    const wrapper = getWrapper({
      props: {
        modelValue: 4,
        total: 100,
      },
    });
    wrapper.vm.first();
    expect(getCalls<[number]>(wrapper, "update:modelValue")[0][0]).toBe(1);
  });
  it("should set page to last possible page when last is clicked", () => {
    const wrapper = getWrapper({
      props: {
        modelValue: 4,
        total: 100,
      },
    });
    wrapper.vm.last();
    expect(getCalls<[number]>(wrapper, "update:modelValue")[0][0]).toBe(10);
  });
  it("should increment page by one when next is clicked", () => {
    const wrapper = getWrapper({
      props: {
        modelValue: 4,
        total: 100,
      },
    });
    wrapper.vm.next();
    expect(getCalls<[number]>(wrapper, "update:modelValue")[0][0]).toBe(5);
  });
  it("should decrement page by one when previous is clicked", () => {
    const wrapper = getWrapper({
      props: {
        modelValue: 4,
        total: 100,
      },
    });
    wrapper.vm.previous();
    expect(getCalls<[number]>(wrapper, "update:modelValue")[0][0]).toBe(3);
  });
  it("should display the correct range text for first and second pages", async () => {
    const wrapper = getWrapper({
      props: {
        modelValue: 1,
        total: 100,
      },
    });
    expect(wrapper.vm.rangeText).toBe("1 - 10 of 100");
    await wrapper.setProps({
      modelValue: 2,
    });
    expect(wrapper.vm.rangeText).toBe("11 - 20 of 100");
  });
  it("should display the correct range text for the last page that is not a round value of page size", () => {
    const wrapper = getWrapper({
      props: {
        modelValue: 11,
        total: 103,
      },
    });
    expect(wrapper.vm.rangeText).toBe("101 - 103 of 103");
  });
});
