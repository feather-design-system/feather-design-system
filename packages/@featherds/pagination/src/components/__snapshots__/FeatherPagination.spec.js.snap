// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatherPagination.vue should disable first and previous buttons on first page 1`] = `
<div
  class="feather-pagination"
>
  <div
    class="per-page-text"
  >
    Rows per page
  </div>
  <feather-select-stub
    class="page-size-select"
    inline=""
    label="Rows per page"
    modelvalue="[object Object]"
    options="[object Object],[object Object],[object Object]"
    textprop="_text"
  />
  <div
    class="range-text"
  >
    1 - 10 of 100
  </div>
  <nav
    aria-label="Pagination controls"
  >
    <ul>
      <li>
        <feather-button-stub
          asanchor="true"
          class="action"
          data-ref-id="feather-pagination-first-button"
          disabled="true"
          href="#"
          icon="Go to first page"
          primary="false"
          secondary="false"
          text="false"
        >
          <feather-icon-stub
            flex="false"
            icon="[object Object]"
          />
        </feather-button-stub>
      </li>
      <li>
        <feather-button-stub
          asanchor="true"
          class="action"
          data-ref-id="feather-pagination-previous-button"
          disabled="true"
          href="#"
          icon="Go to previous page"
          primary="false"
          secondary="false"
          text="false"
        >
          <feather-icon-stub
            flex="false"
            icon="[object Object]"
          />
        </feather-button-stub>
      </li>
      <li>
        <feather-button-stub
          asanchor="true"
          class="action"
          data-ref-id="feather-pagination-next-button"
          disabled="false"
          href="#"
          icon="Go to next page"
          primary="false"
          secondary="false"
          text="false"
        >
          <feather-icon-stub
            flex="false"
            icon="[object Object]"
          />
        </feather-button-stub>
        <feather-button-stub
          asanchor="true"
          class="action"
          data-ref-id="feather-pagination-last-button"
          disabled="false"
          href="#"
          icon="Go to last page"
          primary="false"
          secondary="false"
          text="false"
        >
          <feather-icon-stub
            flex="false"
            icon="[object Object]"
          />
        </feather-button-stub>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`FeatherPagination.vue should disable last and next buttons on last page 1`] = `
<div
  class="feather-pagination"
>
  <div
    class="per-page-text"
  >
    Rows per page
  </div>
  <feather-select-stub
    class="page-size-select"
    inline=""
    label="Rows per page"
    modelvalue="[object Object]"
    options="[object Object],[object Object],[object Object]"
    textprop="_text"
  />
  <div
    class="range-text"
  >
    11 - 20 of 20
  </div>
  <nav
    aria-label="Pagination controls"
  >
    <ul>
      <li>
        <feather-button-stub
          asanchor="true"
          class="action"
          data-ref-id="feather-pagination-first-button"
          disabled="false"
          href="#"
          icon="Go to first page"
          primary="false"
          secondary="false"
          text="false"
        >
          <feather-icon-stub
            flex="false"
            icon="[object Object]"
          />
        </feather-button-stub>
      </li>
      <li>
        <feather-button-stub
          asanchor="true"
          class="action"
          data-ref-id="feather-pagination-previous-button"
          disabled="false"
          href="#"
          icon="Go to previous page"
          primary="false"
          secondary="false"
          text="false"
        >
          <feather-icon-stub
            flex="false"
            icon="[object Object]"
          />
        </feather-button-stub>
      </li>
      <li>
        <feather-button-stub
          asanchor="true"
          class="action"
          data-ref-id="feather-pagination-next-button"
          disabled="true"
          href="#"
          icon="Go to next page"
          primary="false"
          secondary="false"
          text="false"
        >
          <feather-icon-stub
            flex="false"
            icon="[object Object]"
          />
        </feather-button-stub>
        <feather-button-stub
          asanchor="true"
          class="action"
          data-ref-id="feather-pagination-last-button"
          disabled="true"
          href="#"
          icon="Go to last page"
          primary="false"
          secondary="false"
          text="false"
        >
          <feather-icon-stub
            flex="false"
            icon="[object Object]"
          />
        </feather-button-stub>
      </li>
    </ul>
  </nav>
</div>
`;
